### PR TITLE
support binding params with leading :colon

### LINKED ISSUE #22: PDO bindParam should support named parameters with leading colon :
Consider the following statement:

        INSERT INTO `doc`.`tweets` (`id`, `text`, `user`) VALUES (:qp0, :qp1, :qp2)

When I prepare the statement, the value of namedToPositionalParams is:

        array(3) {
        ["qp0"]=>
        int(0)
        ["qp1"]=>
        int(1)
        ["qp2"]=>
        int(2)
        }

When I call:

        $pdoStatement->bindParam(":qp0", $value, $dataType);

My statement fails with:

        Exception\UnsupportedException("0-based parameter binding not supported, use 1-based");

This is because of line 277 of PDOStatement.php:

        if (array_key_exists($parameter, $this->namedToPositionalMap))

The keys in namedToPositionalParams should have leading colons, such as [":qp0"=>0, ":qp1"=>1, ":qp2"=>2].  http://php.net/manual/en/pdostatement.bindparam.php states: Parameter identifier. For a prepared statement using named placeholders, this will be a parameter name of the form :name

This can be fixed by moving the pattern's capturing group to include the colon on line 109 of PDOStatement.php:

        $pattern = '/(:(?:[\w|\d|_](?=([^\'\\\]*(\\\.|\'([^\'\\\]*\\\.)*[^\'\\\]*\'))*[^\']*$))*)/';

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/16059652-pdo-bindparam-should-support-named-parameters-with-leading-colon?utm_campaign=plugin&utm_content=tracker%2F5517944&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F5517944&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by caleblloyd**
Fixes #22 while maintaining backwards compatibility

### CODE REVIEW

**Comment by chaudum** [comment]
hi @caleblloyd 

looks good to me.

can you please add an entry in the `CHANGES.txt` ("Unreleased" section) about your fix and squash the commits.

ps: have you signed a CLA already? you can find them here: https://crate.io/community/contribute/agreements/

File: src/Crate/PDO/PDOStatement.php [hunk 1]
```diff
@@ -274,13 +274,20 @@ public function bindParam(
...
-        if (array_key_exists($parameter, $this->namedToPositionalMap)) {
-            $parameter = $this->namedToPositionalMap[$parameter];
-        } else {
+        if (is_numeric($parameter)) {
             if ($parameter == 0) {
                 throw new Exception\UnsupportedException("0-based parameter binding not supported, use 1-based");
             }
             $parameter--;
+        } else {
+            $namedParameterKey = substr($parameter, 0, 1) == ':' ? substr($parameter, 1) : $parameter;
```
Comment (Ocramius) [inline src/Crate/PDO/PDOStatement.php]: `===`

Comment (caleblloyd) [inline src/Crate/PDO/PDOStatement.php]: it's substr, not strpos, shouldn't make a difference

Comment (Ocramius) [inline src/Crate/PDO/PDOStatement.php]: @caleblloyd it's a general suggestion - would be picked up by typical IDE CS checks as well

**Comment by caleblloyd** [comment]
- Signed CLA
- Added CHANGES.txt entry
- Squashed Commits

Think we're ready to ship!

**Comment by caleblloyd** [comment]
@Ocramius got it, added the ===

**Comment by chaudum** [comment]
@caleblloyd 
cool, we're good to go! unfortunately we haven't received the CLA yet. I will investigate and merge as soon as we have it :)

**Comment by chaudum** [comment]
@caleblloyd thanks for the contribution!

**Comment by caleblloyd** [comment]
@chaudum sure thing, so far with this fix crate has been doing an excellent job of working Yii Framework's ORM.  I'm really enjoying using crate!

**Comment by caleblloyd** [comment]
any plans to publish a release soon so Composer will pick this up?

**Comment by chaudum** [comment]
@caleblloyd thanks for reminding! we'll do a release asap

**Comment by chaudum** [comment]
@caleblloyd 
0.3.0 is now available from packagist https://packagist.org/packages/crate/crate-pdo

**Pull request merged**